### PR TITLE
NPM google-spreadsheet typings

### PR DESCRIPTION
**Pull request by the-vampiire**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by taufn** [comment]
Looking forward to having this types to be merged!

**Review by peterblazejewicz — APPROVED** [review]
Can you folks add a basic set of tests? This could be something really as quick a grabing a code like that one from project readme?
```ts
/// <reference types="node" />

import { GoogleSpreadsheet } from 'google-spreadsheet';

(async () => {
    const doc = new GoogleSpreadsheet('<the sheet ID from the url>');
    await doc.useServiceAccountAuth({
        client_email: process.env.GOOGLE_SERVICE_ACCOUNT_EMAIL!,
        private_key: process.env.GOOGLE_PRIVATE_KEY!,
    });
    await doc.useServiceAccountAuth({
        client_email: '',
        private_key: '',
    });
    // OR use API key -- only for read-only access to public sheets
    doc.useApiKey('YOUR-API-KEY');

    await doc.loadInfo(); // loads document properties and worksheets
    console.log(doc.title);
    await doc.updateProperties({ title: 'renamed doc' });

    const sheet = doc.sheetsByIndex[0]; // or use doc.sheetsById[id]
    console.log(sheet.title);
    console.log(sheet.rowCount);

    // adding / removing sheets
    const newSheet = await doc.addSheet({ title: 'hot new sheet!' });
    await newSheet.delete();
})();

(async () => {
    const doc = new GoogleSpreadsheet('<the sheet ID from the url>');
    // create a sheet and set the header row
    const sheet = await doc.addSheet({ headerValues: ['name', 'email'] });

    // append rows
    const larryRow = await sheet.addRow({ name: 'Larry Page', email: 'larry@google.com' });
    const moreRows = await sheet.addRows([
        { name: 'Sergey Brin', email: 'sergey@google.com' },
        { name: 'Eric Schmidt', email: 'eric@google.com' },
    ]);

    // read rows
    const rows = await sheet.getRows(); // can pass in { limit, offset }

    // read/write row values
    console.log(rows[0].name); // 'Larry Page'
    rows[1].email = 'sergey@abc.xyz'; // update a value
    await rows[1].save(); // save updates
    await rows[1].delete(); // delete a row
})();

(async () => {
    const doc = new GoogleSpreadsheet('<the sheet ID from the url>');
    // create a sheet and set the header row
    const sheet = await doc.addSheet({ headerValues: ['name', 'email'] });
    await sheet.loadCells('A1:E10'); // loads a range of cells
    console.log(sheet.cellStats); // total cells, loaded, how many non-empty
    const a1 = sheet.getCell(0, 0); // access cells using a zero-based index
    const c6 = sheet.getCellByA1('C6'); // or A1 style notation
    // access everything about the cell
    console.log(a1.value);
    console.log(a1.formula);
    console.log(a1.formattedValue);
    // update the cell contents and formatting
    a1.value = 123.456;
    c6.formula = '=A1';
    a1.textFormat = { bold: true };
    c6.note = 'This is a note!';
    await sheet.saveUpdatedCells(); // save all updates in one call
})();
```
(you would see compilation error when doing this, some options are now required in the API? This is why the tests are welcome to verify/confirm API shape)

**Review by peterblazejewicz — COMMENTED** [review]
~~duplicate content redacted~~

**Review by peterblazejewicz — CHANGES_REQUESTED** [review]
ops, sorry for an accidental approval, I meant to use 'request changes' :(

File: types/google-spreadsheet/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        "baseUrl": "../",
+        "typeRoots": [
+            "../"
+        ],
+        "types": [],
+        "noEmit": true,
+        "forceConsistentCasingInFileNames": true
+    },
+    "files": [
+        "index.d.ts"
```
Comment (peterblazejewicz) [inline types/google-spreadsheet/tsconfig.json:20]: Please add a test file here
```suggestion
        "index.d.ts",
        "google-spreadsheet-tests.d.ts"
```

**Comment by EndBug** [comment]
@peterblazejewicz Thanks you for the review! 
We're having an issue with the tests, since nothing can be done without authenticating: is there a way to add tokens for the tests? And if there's not, how could we try to add them?

**Comment by peterblazejewicz** [comment]
> We're having an issue with the tests, since nothing can be done without authenticating: is there a way to add tokens for the tests?

Those tests are run only with the DT linter/compiler, so they are not executed (no requests are ever done):
https://github.com/definitelyTyped/DefinitelyTyped/#create-a-new-package
> This contains sample code which tests the typings. This code does not run, but it is type-checked.

Tests are here to verify types, e.g. againsts the common usage scenarios, interface correctness, etc.

With your specific definition there is a conflict between sample taken from the project README and the contract for the interface in DT, e.g. for an update. The tests will answer the question if the interface should be changed to allow nullable properies (e.g. to update only`title`)

Otherwise great work :)

File: types/google-spreadsheet/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        "noImplicitThis": true,
+        "strictFunctionTypes": true,
+        "strictNullChecks": true,
+        "baseUrl": "../",
+        "typeRoots": ["../"],
+        "types": [],
+        "noEmit": true,
+        "forceConsistentCasingInFileNames": true
+    },
+    "files": ["index.d.ts", "google-spreadsheet-tests.d.ts"]
```
Comment (peterblazejewicz) [inline types/google-spreadsheet/tsconfig.json:15]:
```suggestion
    "files": ["index.d.ts", "google-spreadsheet-tests.ts"]
```

File: types/google-spreadsheet/google-spreadsheet-tests.d.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import { GoogleSpreadsheet } from 'google-spreadsheet';
```
Comment (peterblazejewicz) [inline types/google-spreadsheet/google-spreadsheet-tests.d.ts:1]: Please rename this file to `google-spreadsheet-tests.ts` (drop `.d` from filename, as mentioned in `tsconfig.json` hint

File: types/google-spreadsheet/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,1154 @@
...
+     * - insert:?DESCRIPTION?
+     */
+    addRow(
+        values:
+            | {
+                  [header: string]: string | number | boolean;
+              }
+            | Array<string | number | boolean>,
+        options?: { raw: boolean; insert: boolean },
+    ): Promise<void>;
```
Comment (peterblazejewicz) [inline types/google-spreadsheet/index.d.ts:785]: There is a clash with the original source file and return here:
https://github.com/theoephraim/node-google-spreadsheet/blob/abb8217c2d9b6fc1da731ec9dfeda3f97b349747/lib/GoogleSpreadsheetWorksheet.js#L403-L405
The method cannot be Promise<void>, as it returns `rows[0]`, probably:
```suggestion
    ): Promise<GoogleSpreadsheetRow>;
```

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
minor correction for `addRow` and good to go!

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@the-vampiire @EndBug 
thx for the contribution!

**Comment by EndBug** [comment]
@peterblazejewicz Thank you for the help :)

**Comment by denisatfn** [comment]
This is still broken, getting property `addRow` does not exist on type `GoogleSpreadsheet`

**Comment by EndBug** [comment]
@denisatfn If there's an issue it may be better if you either open a new issue or, even better, open a PR with the desired change. This way only who needs to be pinged will be notified.

**Pull request merged**